File: UPCODER/Other/Luythuatb.py
def binaryMul(a, b, m):
    res = 0
    a = a % m
    while b > 0:
        if b % 2 != 0:
            res = (res + a) % m
        a = (a << 1) % m
        b //= 2
    return res

def binaryPow(a, b, m):
    res = 1
    a = a % m
    while b > 0:
        if b % 2 != 0:
            res = binaryMul(res, a, m)
        a = binaryMul(a, a, m)
        b //= 2
    return res
    
a, b = map(int, input().split())
print(binaryPow(a, b, 1000000007))

File: UPCODER/Other/Luythuatb.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
const ll MOD = 1e9 + 7;

ll binaryMul(ll a, ll b, ll MOD) {
    ll res = 0;
    a = a % MOD;
    
    while (b) {
        if (b & 1) res = (res + a) % MOD;
        a = (a << 1) % MOD;
        b /= 2;
    }
    
    return res;
}

ll binaryPow(ll a, ll b, ll MOD) {
    ll res = 1;
    a = a % MOD;
    
    while (b) {
        if (b & 1) res = binaryMul(res, a, MOD);
        a = binaryMul(a, a, MOD);
        b /= 2;
    }
    
    return res;
}

int main() {
    ll a, b;
    cin >> a >> b;
    cout << binaryPow(a, b, MOD);
    return 0;
}